### PR TITLE
doc: Add support for west in zephyr application extension

### PR DESCRIPTION
**Pull request by carlescufi**
Sample output:

CMake only:
```
# On Linux/macOS
cd samples/hello_world
mkdir build && cd build

# On Windows
cd samples\hello_world
mkdir build & cd build

# Use cmake to configure a Make-based build system:
cmake -DBOARD=native_posix -DSHIELD=x_nucleo_iks01a1 -DCONF_FILE=a.conf b.conf VERBOSE=1 ..

# Now run make on the generated build system:
make
```

West only, simple:
```
west build -b native_posix
```

West only, with flashing:
```
west build -b native_posix -s samples/hello_world
west flash
```
```
west build -b native_posix -s samples/hello_world -d build/board
west flash -d build/board
```
West only, with CMake args:
```
west build -b native_posix -s samples/hello_world -- -DSHIELD=x_nucleo_iks01a1 -DCONF_FILE=a.conf VERBOSE=1
```

CMake and west:
image

### CODE REVIEW

File: doc/extensions/zephyr/application.py [hunk 1]
```diff
@@ -144,61 +155,105 @@ def run(self):
...
+        build_dir = kwargs['build_dir']
+        kwargs['board'] = None
+        cmake_args = self._cmake_args(**kwargs)
+        cmake_args = ' --{}'.format(cmake_args) if cmake_args != '' else ''
+        # ignore zephyr_app since west needs to run within
+        # the installation. Instead rely on relative path.
+        src = ' -s {}'.format(cd_to) if cd_to else ''
+        dst = ' -d {}'.format(build_dir) if build_dir != 'build' else ''
+
+        if 'build' in goals:
```
Comment (dbkinder) [inline doc/extensions/zephyr/application.py]: Ah, hello_world only has a :run: goal, explaining why there's no output for 'west'.  A quick scan shows there are 39 .rst files using this directive with only a :run: or :menuconfig: goal that would generate no west output:
```
samples/drivers/counter/alarm/README.rst:    :goals: run
samples/net/wpanusb/README.rst:   :goals: run
samples/net/wpanusb/README.rst:   :goals: run
samples/net/sockets/echo_client/README.rst:   :goals: run
samples/net/sockets/echo_client/README.rst:   :goals: run
samples/net/sockets/echo_server/README.rst:   :goals: run
samples/net/sockets/echo_server/README.rst:   :goals: run
samples/net/dhcpv4_client/README.rst:   :goals: run
samples/net/lwm2m_client/README.rst:   :goals: run
samples/net/lwm2m_client/README.rst:   :goals: run
samples/net/telnet/README.rst:   :goals: run
samples/hello_world/README.rst:   :goals: run
samples/hello_world/README.rst:   :goals: run
samples/philosophers/README.rst:   :goals: run
samples/subsys/console/getchar/README.rst:   :goals: run
samples/subsys/console/getline/README.rst:   :goals: run
samples/subsys/logging/logger/README.rst:   :goals: run
samples/mpu/mpu_test/README.rst:   :goals: run
samples/portability/cmsis_rtos_v1/timer_synchronization/README.rst:   :goals: run
samples/portability/cmsis_rtos_v1/philosophers/README.rst:   :goals: run
samples/portability/cmsis_rtos_v2/timer_synchronization/README.rst:   :goals: run
samples/portability/cmsis_rtos_v2/philosophers/README.rst:   :goals: run
samples/boards/96b_argonkey/microphone/README.rst:    :goals: run
samples/boards/96b_argonkey/sensors/README.rst:    :goals: run
samples/synchronization/README.rst:   :goals: run
samples/display/cfb_shell/README.rst:   :goals: run
doc/guides/bluetooth/index.rst:   :goals: run
doc/guides/networking/qemu_setup.rst:   :goals: run
doc/guides/networking/native_posix_setup.rst:   :goals: run
boards/x86/qemu_x86/doc/index.rst:   :goals: run
boards/arm/qemu_cortex_m3/doc/index.rst:   :goals: run
boards/arc/em_starterkit/doc/index.rst:   :goals: menuconfig
boards/arc/iotdk/doc/index.rst:   :goals: menuconfig
boards/nios2/qemu_nios2/doc/index.rst:   :goals: run
boards/riscv32/qemu_riscv32/doc/index.rst:   :goals: run
boards/xtensa/xt-sim/doc/index.rst:   :goals: menuconfig
boards/xtensa/xt-sim/doc/index.rst:   :goals: run
boards/xtensa/qemu_xtensa/doc/index.rst:   :goals: run
tests/kernel/mem_protect/protection/README.rst:   :goals: run
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
Some uses of the  zephyr-app-commands directive won't generate any west commands (though there are cmake commands generated).

File: doc/extensions/zephyr/application.py [hunk 2]
```diff
@@ -28,6 +28,10 @@ class ZephyrAppCommandsDirective(Directive):
 
     Directive options:
 
+    \:tool:
+      which tool to use. Valid options are currently 'cmake', 'west' and 'all'.
+      The default is 'cmake'.
```
[the remark this answers is not in the document]
Comment (carlescufi) [reply doc/extensions/zephyr/application.py]: Yep, in the future. But just not right now yet. I will push a commit to enable use of the new directive in at least one place.

File: doc/extensions/zephyr/application.py (hunk 1 above)
Comment (carlescufi) [reply doc/extensions/zephyr/application.py]: Thanks for catching this, will fix

**Comment by carlescufi** [comment]
@dbkinder Fixed the issue you found, along with improving the handling of the different goals. Thanks!

File: doc/extensions/zephyr/application.py [hunk 3]
```diff
@@ -112,6 +118,7 @@ def run(self):
 
         # Parse directive options.  Don't use os.path.sep or os.path.join here!
         # That would break if building the docs on Windows.
+        tool = self.options.get('tool', 'cmake').lower()
```
[the remark this answers is not in the document]
Comment (dbkinder) [reply doc/extensions/zephyr/application.py]: Latest version looks fine:

image

**Review by dbkinder — APPROVED** [review]
LGTM, thanks.

File: doc/extensions/zephyr/application.py [hunk 4]
```diff
@@ -144,61 +155,113 @@ def run(self):
...
+
+    def _generate_west(self, **kwargs):
+        content = []
+        board = kwargs['board']
+        goals = kwargs['goals']
+        cd_to = kwargs['cd_to']
+        build_dir = kwargs['build_dir']
+        kwargs['board'] = None
+        cmake_args = self._cmake_args(**kwargs)
+        cmake_args = ' --{}'.format(cmake_args) if cmake_args != '' else ''
```
Comment (mbolivar) [inline doc/extensions/zephyr/application.py]: Missing space between `--` and `{}`?

**Review by mbolivar — APPROVED** [review]
Looks good, thanks!

File: doc/extensions/zephyr/application.py (hunk 4 above)
Comment (carlescufi) [reply doc/extensions/zephyr/application.py]: No, `cmake_args` has a leading space

**Comment by marc-hb** [comment]
BTW I always found the `mkdir build && cd build` dance very inconvenient, not the least because it means the current directory is a generated and temporary (!) directory. By chance in a (rare) cmake chat with @SebastianBoe I learned the proper and brand new (3.13) -B way to use cmake and my life literally changed:
```shell
cmake -GNinja -DBOARD=qemu_x86 -B build # mkdir comes "for free"
ninja -C build # or make -C build
edit src/main.c # good bye all the '..'s
rm -rf build # good luck doing that with the current documentation.
```
Enough `cd` commands already!

So, after enjoying this wonderful new life for a number of weeks and making sure there was no drawback whatsoever, I decided to be a good open-source citizen and share this better way in the getting started guide. Then I realized these short command samples are generated. Then I found this script that does it. Then I decided that... life is too short to understand this script :-( 
If on the other hand you're one of the few people who already understand this script then you may consider this better way?

Wait... of course west champions won't care a bit about invoking cmake directly; what was I thinking ;-)

**Comment by mbolivar** [comment]
> By chance in a (rare) cmake chat with @SebastianBoe I learned the proper and brand new (3.13) -B way to use cmake and my life literally changed

FWIW, when we first decided to write `west build`, CMake 3.13 was more recent than the minimum  version supported by Zephyr. 

> Then I decided that... life is too short to understand this script :-(

I'm sorry to hear that understanding this script was a barrier to contributing to the documentation.

Again FWIW, having this logic in one place makes it easier to make tree-wide changes to the documentation whenever a better way of accomplishing some build system related task is found, such as the improvement you have pointed out in this comment.

If you don't want this information to get lost but don't want to update the script yourself, please consider opening an issue against the script to update its output when the `tool` is `"cmake"`.

**Comment by marc-hb** [comment]
> If you don't want this information to get lost but don't want to update the script yourself, please consider opening an issue against the script to update its output when the tool is "cmake".

Done, issue #15315

**Pull request merged**